File: java/src/anotherfuncsplayer/Pathfinding.java
package anotherfuncsplayer;

import battlecode.common.*;

import java.util.Random;
public class Pathfinding {
    static RobotController rc;
    public static MapLocation target;
    static boolean[] impassable = null;
    static final Random rng = new Random(6147);
    static boolean changedTarget = false;
    static final Direction[] directions = {
            Direction.NORTH,
            Direction.NORTHEAST,
            Direction.EAST,
            Direction.SOUTHEAST,
            Direction.SOUTH,
            Direction.SOUTHWEST,
            Direction.WEST,
            Direction.NORTHWEST,
            Direction.CENTER
    };
    public static void init(RobotController r) {
        rc = r;
        anotherfuncsplayer.Util.init(rc);
        
        BugNav.rotateRight = Util.rng.nextDouble() > 0.5;
    }
    
    static void setImpassable(boolean[] imp) {
        impassable = imp;
    }
    public static void initTurn() {
        impassable = new boolean[directions.length];
    }
    static boolean canMove(Direction dir) {
        if (!rc.canMove(dir))
            return false;
        if (impassable[dir.ordinal()])
            return false;
        return true;
    }
    
    static public void paint(boolean isSplasher) throws GameActionException {
    	MapInfo[] nearbyTiles = rc.senseNearbyMapInfos();
    	MapLocation closest_srp = new MapLocation(rc.getLocation().x - (rc.getLocation().x % 4) + 2, rc.getLocation().y - (rc.getLocation().y % 4) + 2);
		if (rc.canCompleteResourcePattern(closest_srp)) {
			rc.completeResourcePattern(closest_srp);
		}
    	
    	if (isSplasher) {
    		boolean existsEmpty = false;
            for (MapInfo tile : nearbyTiles) {
            	if (tile.getPaint() == PaintType.ALLY_SECONDARY) {
            		existsEmpty = false;
            	}
            	else if (tile.getPaint() == PaintType.EMPTY || tile.getPaint().isEnemy()) {
            		existsEmpty = true;
            	}
            }
            if (!existsEmpty) {
        		return ;
            }
    	}
    	else if (rc.getType() == UnitType.MOPPER) {
    		MapInfo currentTile = rc.senseMapInfo(rc.getLocation());
            if (currentTile.getPaint().isEnemy() && rc.canAttack(rc.getLocation())){
            	rc.attack(rc.getLocation());
        	}
            RobotInfo[] nearbyRobots = rc.senseNearbyRobots();
            for (RobotInfo robo : nearbyRobots){
            	if (robo.team != rc.getTeam() && (robo.type == UnitType.SOLDIER || robo.type == UnitType.SPLASHER)) {
	                MapLocation my_target = robo.location;
	                Direction dir = rc.getLocation().directionTo(my_target);
	                if (rc.canMopSwing(dir) && robo.team!=rc.getTeam()){
	                    rc.mopSwing(dir);
	                    return ;
	                }
            	}
            	if (robo.team == rc.getTeam() && robo.type == UnitType.SOLDIER) {
            		if (robo.paintAmount <= 75 && rc.getPaint() >= 50 && rc.canTransferPaint(robo.location, 25)) {
            			rc.transferPaint(robo.location, 25);
            			return ;
            		}
            	}
            }
            return ;
    	}
    	MapInfo currentTile = rc.senseMapInfo(rc.getLocation());
        if (!currentTile.getPaint().isAlly() && rc.canAttack(rc.getLocation())){
        	rc.attack(rc.getLocation());
    	}
    }
    
    static public void setTarget(MapLocation newTarget) {
    	target = newTarget;
    }
    
    static public void splasherNewTarget(MapLocation target) throws GameActionException {
    	MapLocation rotation = new MapLocation(target.y, target.x);
    	rotation = Util.clipToWithinMap(rotation);
    	setTarget(rotation);
    }
    
    static public void move(MapLocation loc, boolean isSplasher) throws GameActionException {
    	target = loc;
    	BugNav.init(rc);
    	BugNav.initTurn();
        if (!BugNav.nav(target)) {
        	greedyPath();
        	paint(isSplasher);
        	
        }
		else {
			BugNav.move(target);
			paint(isSplasher);
        }
    }
    
    static final double eps = 1e-5;
    static void greedyPath() {
        try {
            MapLocation myLoc = rc.getLocation();
            Direction bestDir = null;
            double bestEstimation = 0;
            int bestEstimationDist = 0;
            for (Direction dir : directions) {
                MapLocation newLoc = myLoc.add(dir);
                if (!rc.onTheMap(newLoc))
                    continue;
                if (!canMove(dir))
                    continue;
                if (!strictlyCloser(newLoc, myLoc, target))
                    continue;
                int newDist = newLoc.distanceSquaredTo(target);
                // TODO: Better estimation?
                double estimation = 1 + Util.distance(target, newLoc);
                if (bestDir == null || estimation < bestEstimation - eps
                        || (Math.abs(estimation - bestEstimation) <= 2 * eps && newDist < bestEstimationDist)) {
                    bestEstimation = estimation;
                    bestDir = dir;
                    bestEstimationDist = newDist;
                }
            }
            if (bestDir != null) {
                rc.move(bestDir);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    static boolean strictlyCloser(MapLocation newLoc, MapLocation oldLoc, MapLocation target) {
        int dOld = Util.distance(target, oldLoc), dNew = Util.distance(target, newLoc);
        if (dOld < dNew)
            return false;
        if (dNew < dOld)
            return true;
        return target.distanceSquaredTo(newLoc) < target.distanceSquaredTo(oldLoc);
    }
}

File: python/src/translated/Pathfinding.py
import random
from battlecode25.stubs import *

class Pathfinding:
    rc = None
    target = None
    impassable = None
    rng = random.Random(6147)
    changed_target = False
    has_resource = False
    directions = [
        Direction.NORTH,
        Direction.NORTHEAST,
        Direction.EAST,
        Direction.SOUTHEAST,
        Direction.SOUTH,
        Direction.SOUTHWEST,
        Direction.WEST,
        Direction.NORTHWEST,
        Direction.CENTER
    ]

    @staticmethod
    def init(r: RobotController):
        Pathfinding.rc = r
        Util.init(r)
        BugNav.rotate_right = Util.rng.random() > 0.5
        if r.getMapWidth() >= 30 or r.getMapHeight() >= 30:
            Pathfinding.has_resource = True

    @staticmethod
    def set_impassable(imp):
        Pathfinding.impassable = imp

    @staticmethod
    def init_turn():
        Pathfinding.impassable = [False] * len(Pathfinding.directions)

    @staticmethod
    def can_move(dir: Direction) -> bool:
        if not Pathfinding.rc.canMove(dir):
            return False
        if Pathfinding.impassable[dir.ordinal()]:
            return False
        return True

    @staticmethod
    def paint(is_splasher: bool):
        if is_splasher:
            exists_empty = False
            nearby_tiles = Pathfinding.rc.senseNearbyMapInfos()
            for tile in nearby_tiles:
                if tile.hasRuin() and Pathfinding.rc.senseRobotAtLocation(tile.getMapLocation()) is not None and Pathfinding.rc.senseRobotAtLocation(tile.getMapLocation()).team != Pathfinding.rc.getTeam():
                    Pathfinding.set_target(tile.getMapLocation())
                    if Pathfinding.rc.canAttack(Pathfinding.target):
                        Pathfinding.rc.attack(tile.getMapLocation())
                elif Pathfinding.has_resource and tile.getMark().isAlly():
                    exists_empty = False
                elif tile.getPaint() == PaintType.EMPTY or tile.getPaint().isEnemy():
                    exists_empty = True
            if not exists_empty:
                return

        current_tile = Pathfinding.rc.senseMapInfo(Pathfinding.rc.getLocation())
        if not current_tile.getPaint().isAlly() and Pathfinding.rc.canAttack(Pathfinding.rc.getLocation()):
            Pathfinding.rc.attack(Pathfinding.rc.getLocation())

    @staticmethod
    def set_target(new_target: MapLocation):
        Pathfinding.target = new_target

    @staticmethod
    def splasher_new_target(target: MapLocation):
        rotation = MapLocation(target.y, target.x)
        rotation = Util.clip_to_within_map(rotation)
        Pathfinding.set_target(rotation)

    @staticmethod
    def move(loc: MapLocation, is_splasher: bool):
        Pathfinding.target = loc
        BugNav.init(Pathfinding.rc)
        BugNav.init_turn()
        if not BugNav.nav(Pathfinding.target):
            Pathfinding.greedy_path()
            Pathfinding.paint(is_splasher)
        else:
            BugNav.move(Pathfinding.target)
            Pathfinding.paint(is_splasher)

    eps = 1e-5

    @staticmethod
    def greedy_path():
        try:
            my_loc = Pathfinding.rc.getLocation()
            best_dir = None
            best_estimation = 0
            best_estimation_dist = 0
            for dir in Pathfinding.directions:
                new_loc = my_loc.add(dir)
                if not Pathfinding.rc.onTheMap(new_loc):
                    continue
                if not Pathfinding.can_move(dir):
                    continue
                if not Pathfinding.strictly_closer(new_loc, my_loc, Pathfinding.target):
                    continue
                new_dist = new_loc.distanceSquaredTo(Pathfinding.target)
                estimation = 1 + Util.distance(Pathfinding.target, new_loc)
                if best_dir is None or estimation < best_estimation - Pathfinding.eps or (abs(estimation - best_estimation) <= 2 * Pathfinding.eps and new_dist < best_estimation_dist):
                    best_estimation = estimation
                    best_dir = dir
                    best_estimation_dist = new_dist
            if best_dir is not None:
                Pathfinding.rc.move(best_dir)
        except Exception as e:
            print(e)

    @staticmethod
    def strictly_closer(new_loc: MapLocation, old_loc: MapLocation, target: MapLocation) -> bool:
        d_old = Util.distance(target, old_loc)
        d_new = Util.distance(target, new_loc)
        if d_old < d_new:
            return False
        if d_new < d_old:
            return True
        return target.distanceSquaredTo(new_loc) < target.distanceSquaredTo(old_loc)
